src/server/worker/pipeline/transform/registry.go

```diff
@@ -331,111 +377,69 @@ func (reg *registry) processDatums(ctx context.Context, pj *pendingJob, taskDoer
+			input, err := serializeCreateDatumSetsTask(&CreateDatumSetsTask{
+				FileSetId: fileSetID,
+				PathRange: shard,
+				SetSpec:   setSpec,
+			})
+			if err != nil {
+				return err
+			}
+			inputs = append(inputs, input)
 		}
-		datumSetsFileSetID = result.FileSetId
-		return nil
-	}); err != nil {
-		return "", errors.EnsureStack(err)
-	}
-	return datumSetsFileSetID, nil
-}
-
-func processDatumSets(pachClient *client.APIClient, pj *pendingJob, taskDoer task.Doer, fileSetID string, stats *datum.Stats) error {
-	return errors.EnsureStack(pj.logger.LogStep("processing datum sets", func() error {
-		eg, ctx := errgroup.WithContext(pachClient.Ctx())
-		pachClient := pachClient.WithCtx(ctx)
-		inputChan := make(chan *types.Any)
-		eg.Go(func() error {
-			defer close(inputChan)
-			commit := client.NewRepo(client.FileSetsRepoName).NewCommit("", fileSetID)
-			r, err := pachClient.GetFileTAR(commit, "/*")
+		return task.DoBatch(pachClient.Ctx(), taskDoer, inputs, func(_ int64, output *types.Any, err error) error {
 			if err != nil {
 				return err
 			}
-			if err := tarutil.Iterate(r, func(f tarutil.File) error {
-				buf := &bytes.Buffer{}
-				if err := f.Content(buf); err != nil {
-					return errors.EnsureStack(err)
-				}
-				input := &types.Any{}
-				if err := proto.Unmarshal(buf.Bytes(), input); err != nil {
-					return errors.EnsureStack(err)
-				}
-				select {
-				case inputChan <- input:
-				case <-ctx.Done():
-					return errors.EnsureStack(ctx.Err())
-				}
-				return nil
-			}, true); err != nil && !pfsserver.IsFileNotFoundErr(err) {
+			result, err := deserializeCreateDatumSetsTaskResult(output)
+			if err != nil {
 				return err
 			}
+			datumSets = append(datumSets, result.DatumSets...)
 			return nil
 		})
-		eg.Go(func() error {
-			return errors.EnsureStack(taskDoer.Do(
-				ctx,
-				inputChan,
-				func(_ int64, output *types.Any, err error) error {
-					if err != nil {
-						return err
-					}
-					data, err := deserializeDatumSet(output)
-					if err != nil {
-						return err
-					}
-					if _, err := pachClient.PfsAPIClient.AddFileSet(
-						pachClient.Ctx(),
-						&pfs.AddFileSetRequest{
-							Commit:    pj.commitInfo.Commit,
-							FileSetId: data.OutputFileSetId,
-						},
-					); err != nil {
-						return grpcutil.ScrubGRPC(err)
-					}
-					if _, err := pachClient.PfsAPIClient.AddFileSet(
-						pachClient.Ctx(),
-						&pfs.AddFileSetRequest{
-							Commit:    pj.metaCommitInfo.Commit,
-							FileSetId: data.MetaFileSetId,
-						},
-					); err != nil {
-						return grpcutil.ScrubGRPC(err)
-					}
-					if err := datum.MergeStats(stats, data.Stats); err != nil {
-						return err
-					}
-					pj.saveJobStats(data.Stats)
-					return pj.writeJobInfo()
-				},
-			))
-		})
-		return errors.EnsureStack(eg.Wait())
-	}))
+	}); err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+	return datumSets, nil
+}
+
+func createSetSpec(pj *pendingJob) (*datum.SetSpec, error) {
```

Comment:
nit: I think this belongs in pending_job.go

Reply:
I think there is some refactoring that will come with removing parallel jobs which will satisfy this. I actually think we might just get rid of the notion of a pending job since there won't be multiple running at the same time. For now, the datum set creation is managed by the registry, which is why this is here.